x/ibc-hooks/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -0,0 +1,271 @@
+    if let Some(memo) = &next_memo {
+        validate_memo(memo)?;
+    }
+
+    // Check that there isn't anything stored in SWAP_REPLY_STATES. If there is,
+    // it means that the contract is already waiting for a reply and should not
+    // override the stored state. This should only happen if a contract we call
+    // calls back to this one. This is likely a malicious attempt modify the
+    // contract's state before it has replied.
+    if SWAP_REPLY_STATES.may_load(deps.storage)?.is_some() {
+        return Err(ContractError::ContractLocked {
+            msg: "Already waiting for a reply".to_string(),
+        });
+    }
+    // Store information about the original message to be used in the reply
+    SWAP_REPLY_STATES.save(
+        deps.storage,
+        &SwapMsgReplyState {
+            swap_msg,
+            block_time,
+            contract_addr,
+            forward_to: ForwardTo {
+                channel: valid_channel,
+                receiver: valid_receiver,
+                next_memo,
+                failed_delivery,
+            },
+        },
+    )?;
+
+    Ok(Response::new().add_submessage(SubMsg::reply_on_success(msg, SWAP_REPLY_ID)))
+}
+
+pub fn handle_swap_reply(deps: DepsMut, msg: Reply) -> Result<Response, ContractError> {
+    deps.api.debug(&format!("handle_swap_reply"));
+    let swap_msg_state = SWAP_REPLY_STATES.load(deps.storage)?;
+    SWAP_REPLY_STATES.remove(deps.storage);
+
+    // If the swaprouter swap failed, return an error
+    let SubMsgResult::Ok(SubMsgResponse { data: Some(b), .. }) = msg.result else {
+        return Err(ContractError::FailedSwap {
+            msg: format!("No data"),
+        })
+    };
+
+    // Parse underlying response from the chain
+    let parsed =
+        cw_utils::parse_execute_response_data(&b).map_err(|e| ContractError::FailedSwap {
+            msg: format!("failed to parse: {e}"),
+        })?;
+    let swap_response: SwapResponse = from_binary(&parsed.data.unwrap_or_default())?;
+
+    // Build an IBC packet to forward the swap.
+    let contract_addr = &swap_msg_state.contract_addr;
+    let ts = swap_msg_state.block_time.plus_seconds(PACKET_LIFETIME);
+    let config = CONFIG.load(deps.storage)?;
+
+    // If the memo is provided we want to include it in the IBC message
+    let memo: serde_cw_value::Value = if let Some(memo) = &swap_msg_state.forward_to.next_memo {
+        serde_json_wasm::from_str(&memo.to_string()).map_err(|_e| ContractError::InvalidMemo {
+            error: format!("this should be unreachable"),
+            memo: memo.to_string(),
+        })?
+    } else {
+        serde_json_wasm::from_str("{}").unwrap()
+    };
+
+    // If tracking callbacks, we want to include the callback key in the memo
+    // without otherwise modifying the provided one
+    let memo = match config.track_ibc_callbacks {
+        true => {
+            let serde_cw_value::Value::Map(mut m) = memo else { unreachable!() };
```

Comment:
Why don't we just define memo to be of the correct type on line 104?

Reply:
Since Map is a variant, I would have to declare it using the variant's type: `BTreeMap<Value, Value>`, and then deal with more wrappings/unwrappings, I think. 

The only reason this is unreachable is because we run `validate_memo(memo)?;` on line 50 before the reply executes. I can look into passing the btree instead, but it doesn't seem as straightforward when I looked quickly now